Web Inspector: only `awaitPromise` for builtin `@Promise`

dcrousso opened the pull request:
#### a7d66e60ea7f24d6d2ec35a47bb156c050d107ff
<pre>
Web Inspector: only `awaitPromise` for builtin `@Promise`
<a href="https://bugs.webkit.org/show_bug.cgi?id=264920">https://bugs.webkit.org/show_bug.cgi?id=264920</a>

Reviewed by Patrick Angle.

Otherwise our (unchecked) attempts to `.@then()`/`.@catch()` could result in problems. We also don&apos;t really want to support nonstandard behavior (i.e. there&apos;s no guarantee that a custom &quot;thenable&quot; will actually have similar/matching behavior to `Promise.prototype.then`).

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(InjectedScript.prototype.awaitPromise):
(InjectedScript.prototype.callFunctionOn):
(InjectedScript.prototype._forEachPropertyDescriptor.createFakeValueDescriptor):

* LayoutTests/inspector/runtime/awaitPromise.html:
* LayoutTests/inspector/runtime/awaitPromise-expected.txt:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise-expected.txt:

Canonical link: <a href="https://commits.webkit.org/271249@main">https://commits.webkit.org/271249@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00017ab519b70b255b838e4162dc54a3a313e40c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5107 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27739 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28679 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24228 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2515 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24187 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26752 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22749 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3511 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29186 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23087 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24150 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29779 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25702 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3543 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1743 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27673 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4977 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23477 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33151 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6675 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4019 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7185 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

yury-s reviewed (approved):
Neat! I didn't know there is builtin `isPromise`. For my education, where should I look for definitions of methods like this?

patrickangle reviewed (approved):
lgtm, but yeah a line break in the commit message somewhere would be nice 😃

dcrousso committed:
Web Inspector: only `awaitPromise` for builtin `@Promise`
https://bugs.webkit.org/show_bug.cgi?id=264920

Reviewed by Patrick Angle.

Otherwise our (unchecked) attempts to `.@then()`/`.@catch()` could result in problems. We also don't really want to support nonstandard behavior (i.e. there's no guarantee that a custom "thenable" will actually have similar/matching behavior to `Promise.prototype.then`).

* Source/JavaScriptCore/inspector/InjectedScriptSource.js:
(InjectedScript.prototype.awaitPromise):
(InjectedScript.prototype.callFunctionOn):
(InjectedScript.prototype._forEachPropertyDescriptor.createFakeValueDescriptor):

* LayoutTests/inspector/runtime/awaitPromise.html:
* LayoutTests/inspector/runtime/awaitPromise-expected.txt:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise.html:
* LayoutTests/inspector/runtime/callFunctionOn-awaitPromise-expected.txt:

Canonical link: https://commits.webkit.org/271249@main
